src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -367,24 +392,128 @@ private static String getClassNameFromQualifiedName(String qualifiedName) {
+             ++index) {
+            if (index == minLength - 1
+                && shouldCountBeUpdatedAtLastCharacter
+                || patternTypeDeclaration.charAt(index) == packageSeparator) {
+                result = index;
+            }
+        }
+        return result;
+    }
 
-        /** Corresponding node. */
-        private final DetailAST classAst;
+    /**
+     * Maintains information about the type of declaration.
+     * Any ast node of type {@link TokenTypes#CLASS_DEF} or {@link TokenTypes#INTERFACE_DEF}
+     * or {@link TokenTypes#ENUM_DEF} or {@link TokenTypes#ANNOTATION_DEF}
+     * or {@link TokenTypes#RECORD_DEF} is considered as a type declaration.
+     */
+    private static class TypeDeclarationDescription {
 
-        /** Qualified class name(with package). */
+        /**
+         * Complete type declaration name with package name and outer type declaration name.
+         */
         private final String qualifiedName;
 
+        /**
+         * Depth of nesting of type declaration. The depth of top-level class is 0,
+         * a class nested under it has a depth of -1 and so on.
+         */
+        private final int depth;
+
+        /**
+         * Type declaration ast node.
+         */
+        private final DetailAST typeDeclarationAst;
+
+        /**
+         * Create an instance of TypeDeclarationDescription.
+         *
+         * @param qualifiedName Complete type declaration name with package name and outer type
+         *                      declaration name.
+         * @param depth Depth of nesting of type declaration
+         * @param typeDeclarationAst Type declaration ast node
+         */
+        /* package */ TypeDeclarationDescription(String qualifiedName, int depth,
+                                          DetailAST typeDeclarationAst) {
+            this.qualifiedName = qualifiedName;
+            this.depth = depth;
+            this.typeDeclarationAst = typeDeclarationAst;
+        }
+
+        /**
+         * Get the complete type declaration name i.e. type declaration name with package name
+         * and outer type declaration name.
+         *
+         * @return qualified class name
+         */
+        protected String getQualifiedName() {
+            return qualifiedName;
+        }
+
+        /**
+         * Get the depth of type declaration.
+         *
+         * @return the depth of nesting of type declaration
+         */
+        protected int getDepth() {
+            return depth;
+        }
+
+        /**
+         * Get the type declaration ast node.
+         *
+         * @return ast node of the type declaration
+         */
+        protected DetailAST getTypeDeclarationAst() {
+            return typeDeclarationAst;
+        }
+    }
+
+    /**
+     * Maintains information about class's ctors. {@link TypeDeclarationDescription} does not
+     * maintain information about classes, a subclass called {@link ClassDesc}
```

Comment:
It seems odd to have this class to have a link to itself and refer to itself in the "third person".

Reply:
I moved that point to `TypeDeclarationDescription`, done.

Reply:
Resolved.